### DISCUSSION TITLE
Ability to set backend as pytest option

### DISCUSSION THREAD

**Discussion by aaronzo** [networkx/networkx#6819] [General]
It's annoying to reproduce test errors with `dispatch` developing on Windows, since
```bash
NETWORKX_GRAPH_CONVERT=nx-loopback pytest
```
does not execute in powershell/cmd. Of course I can use Git Bash or WSL, but I think the developer experience could be made easier by adding an option in `pytest_addoption` to set the backend.

Something like
```bash
pytest --backend=nx-loopback .
```
would be ideal

**Comment by dschult** [comment]
Can you get this set up and working on your windows machine?

**Reply by aaronzo** [reply]
I've got somewhere by using `pytest_configure`, but the problem seems to be that the functions are registered to dispatch as soon as networkx is imported, and I can't seem to re-register them once I set `os.environ['NETWORKX_GRAPH_CONVERT']`

**Reply by aaronzo** [reply]
In `conftest.py` I added:
```python
import os
import warnings

import pytest

def pytest_addoption(parser):
    parser.addoption(
        "--runslow", action="store_true", default=False, help="run slow tests"
    )
    parser.addoption(
        "--backend", action="store", default="", help="backend to run tests with"
    )


def pytest_configure(config):
    config.addinivalue_line("markers", "slow: mark test as slow to run")
    if (backend := config.getoption("--backend")):
        print(backend)
        os.environ["NETWORKX_GRAPH_CONVERT"] = backend
```

But it seems all the algorithms are already registered before pytest takes control because this option didn't allow me to use the loopback

The only way I can get this to work is adding an argument parser to `utils/backends.py`:
```python
# Check for auto-convert testing
# This allows existing NetworkX tests to be run against a backend
# implementation without any changes to the testing code. The only
# required change is to set an environment variable prior to running
# pytest.
_parser = argparse.ArgumentParser()
_parser.add_argument(
    "--backend", action="store", default="", help="backend to run tests with"
)
_args, _ = _parser.parse_known_args(sys.argv)
if _args.backend or os.environ.get("NETWORKX_GRAPH_CONVERT"):
    plugin_name = _args.backend or os.environ["NETWORKX_GRAPH_CONVERT"]
    if not plugins:
        raise Exception("No registered networkx.plugins entry_points")
    if plugin_name not in plugins:
        raise Exception(
            f"No registered networkx.plugins entry_point named {plugin_name}"
        )

    try:
        import pytest
    except ImportError:
        raise ImportError(
            f"Missing pytest, which is required when using NETWORKX_GRAPH_CONVERT"
        )

    # Override `dispatch` for testing
    _dispatch = test_override_dispatch

    # Whether to call the networkx algorithm if the backend doesn't implement a function
    fallback_to_nx = bool(os.environ.get("NETWORKX_BACKEND_TEST_EXHAUSTIVE"))
```
which doesn't seem ideal

**Comment by aaronzo** [comment]
Okay, I went down a very educational rabbithole with this, but I believe I've got some functionality that might be pretty useful more generally, and avoids the issues in the thread above.

I added a function `reregister_all_algos` (not settled on the name) to `networkx.utils.backends` which re-dispatches and registers every single function, allowing the user to override the dispatch function (most likely with `test_override_dispatch` for testing, without requiring an environment variable to be set.

Would love to hear thoughts @dschult & @eriknw - I've raised a draft PR #6829

**Comment by eriknw** [comment]
Cool. It would be nice to use arguments to pytest instead of environment variables.
